File: Interpreter/CalculateEngine.java
package calculator.xwg;

import java.util.LinkedList;
import java.util.Set;
import android.content.Context;

public class CalculateEngine
{
	final static int CUSTOM_FUN_COUNT = 12;
	FunctionManager functionManager;
	ConstManager constManager = new ConstManager();
	private Context systemContext = null; 
	
	CalculateEngine(Context context){
		systemContext = context;
		functionManager = new FunctionManager(systemContext);
		registerConst();
		registerStandardFunctions();
	}
	
	void registerConst(){
		constManager.registerConst(systemContext.getText(R.string.character_pi).toString(), new Complex(Math.PI));
		constManager.registerConst("e", new Complex(Math.E));
	}
	
	void registerStandardFunctions(){
		functionManager.registerFunction(new AcosFun());
		functionManager.registerFunction(new AcoshFun());
		functionManager.registerFunction(new AsinFun());
		functionManager.registerFunction(new AsinhFun());
		functionManager.registerFunction(new AtanFun());
		functionManager.registerFunction(new AtanhFun());
		functionManager.registerFunction(new AverageFun());
		functionManager.registerFunction(new CosFun());
		functionManager.registerFunction(new CoshFun());
		functionManager.registerFunction(new Log10Fun());
		functionManager.registerFunction(new LogeFun());
		functionManager.registerFunction(new PowerFun());
		functionManager.registerFunction(new RootFun());
		functionManager.registerFunction(new SinFun());
		functionManager.registerFunction(new SinhFun());
		functionManager.registerFunction(new SumFun());
		functionManager.registerFunction(new TanFun());
		functionManager.registerFunction(new FactorialFun());
				
		functionManager.registerFunction(new ConvertFormatFun("to" + systemContext.getText(R.string.character_angle), 
										new RadiusAngleFormatter(systemContext, false)));
		functionManager.registerFunction(new ConvertFormatFun("to" + systemContext.getText(R.string.character_degree), 
										new DegreesFormatter(systemContext)));

		functionManager.registerFunction(new PreDefineFunction("x2", "pow(#1,2)"));
		functionManager.registerFunction(new PreDefineFunction("x3", "pow(#1,3)"));
		functionManager.registerFunction(new PreDefineFunction("2" + systemContext.getText(R.string.character_sqrt), "root(#1,2)"));
		functionManager.registerFunction(new PreDefineFunction("3" + systemContext.getText(R.string.character_sqrt), "root(#1,3)"));
		functionManager.registerFunction(new PreDefineFunction("ex", "pow(e,#1)"));
		
		for(int i = 1; i <= CUSTOM_FUN_COUNT; ++i){ 
			UserDefineFunction udf = UserDefineFunction.load(systemContext, "F" + i);
			if(udf != null){
				functionManager.registerUserDefineFunction(udf);
			}
		}
	}
	
	public LinkedList<Token> analyzeToken(String strQuestion){
		TokenAnalyzer analyzer = new TokenAnalyzer();
	    LinkedList<Token> tokenList = analyzer.analyzeToken(strQuestion, new TokenPatternFactory(){
	    	public int createPatterns(LinkedList<TokenPattern> list)
	        {
	        	String funPattern = PatternBuilder.build(functionManager.functions());
	            
	            if(funPattern.length() > 0){
	            	list.add(new TokenPattern(Token.EType.FunctionName, funPattern));
	            }
	            list.add(new TokenPattern(Token.EType.Parameter, "#[1-9]"));
	            
	            String constPattern = PatternBuilder.build(constManager.consts());
	            list.add(new TokenPattern(Token.EType.Number, "(" + constPattern + ")"));

	            String numberPattern = "(((\\.[0-9]+)|([0-9]+(\\.[0-9]*)?))[eE][+-]?[0-9]+)";
	            numberPattern += "|";
	            numberPattern += "((\\.[0-9]+)|([0-9]+\\.[0-9]*))"; 
	            numberPattern += "|";
	            numberPattern += "([0-9]+)";
	            
	            CharSequence degree = systemContext.getText(R.string.character_degree);
	            list.add(new TokenPattern(Token.EType.Number, "((\\.[0-9]+)|([0-9]+\\.[0-9]*)|([0-9]+))%"));
	            list.add(new TokenPattern(Token.EType.Number, "(" + numberPattern + ")[" + degree + "i]?"));
	            list.add(new TokenPattern(Token.EType.Number, "[i]"));
	            CharSequence angle = systemContext.getText(R.string.character_angle);
	            list.add(new TokenPattern(Token.EType.Operator, "[-+×/" + angle + "]"));
	            list.add(new TokenPattern(Token.EType.Parenthese, "[()]"));
	            list.add(new TokenPattern(Token.EType.Comma, ","));
	            return list.size();
	        }
	    });
	    return tokenList;
	}
	
	public String calculate(String strQuestion, boolean convertFormat){
		if(currentRecord == null){
			currentRecord = new Record();
		}
	    LinkedList<Token> tokenList = analyzeToken(strQuestion);
	    if(!convertFormat){
	    	currentRecord.question = strQuestion;
	    }
	    String result = calculate(tokenList);
	    tokenList = null;
	    return result;
	}

	boolean isFunction(String name){
		return (functionManager.getFunction(name) != null);
	}
	
	CharSequence[] getCustomFunctionItems(){
		CharSequence items[] = new CharSequence[CUSTOM_FUN_COUNT];
		
		for(int i = 1; i <= CUSTOM_FUN_COUNT; ++i){
			String key = "F" + i;
			String item = null;
			
			UserDefineFunction udf = functionManager.getUserDefineFunction(key);
			if(udf != null){
				item = udf.getName() + ":" + udf.getExprString(); 
			}else{
				item = key + ":Empty";
			}
			
			items[i - 1] = item;
		}
		return items;
	}
	
	public int saveCustomFunction(String key, String funName, String funText){
		return functionManager.registerUserDefineFunction(key, funName, funText);
	}
	
	public void clearCustomFunctions(){
		for(int i = 1; i <= CUSTOM_FUN_COUNT; ++i){ 
			UserDefineFunction.clear(systemContext, "F" + i);
		}
	}
	
	public ConstManager getConstManager(){
		return constManager;
	}
	
	CharSequence[] getConstItems(){
		int count = constManager.getConstCount();
		CharSequence items[] = new CharSequence[count];
		
		Set<String> set = constManager.consts();
		
		StandardFormatter formatter = new StandardFormatter(systemContext);
		
		int index = 0;
        for(String key : set)
        {
        	String item = key + ":" + formatter.toString(constManager.find(key));
        	items[index] = item;
        	index++;
        }
        return items;
	}

	public class Record
	{
		boolean success;
	    String question;
	    Complex result;
	}

	public int getRecordCount(){
		return recordList.size();
	}
	
	public Record getRecord(int index){
		if(index >= 0 && index < recordList.size())
	    {
	        return recordList.get(index);
	    }
	    else
	    {
	        return null;
	    }
	}
	
	public boolean saveRecord(){
		if(currentRecord != null && currentRecord.success)
	    {
	        recordList.add(currentRecord);
	        currentRecord = null;
	        return true;
	    }else{
	    	return false;
	    }
	}
	
	CharSequence[] getRecordItems(){
		StandardFormatter formator = new StandardFormatter(systemContext);
		int record_count = getRecordCount();
		CharSequence items[] = new CharSequence[record_count];
		for(int i = 0; i < record_count; ++i){
			Record record = getRecord(i);
			String recordString = formator.toString(record.result) + ":" + record.question;
			items[i] = recordString;
		}
		return items;
	}
	
	public boolean clearRecord(int index){
		if(index >= 0 && index < recordList.size())
	    {
	        recordList.remove(index);
	                                                               
	        return true;
	    }
	    else
	    {
	        return false;
	    }
	}
	
	public boolean clearAllRecord(){
		recordList.clear();
	    return true;
	}
	
	public Context getSystemContext(){
    	return systemContext;
    }
	
	public String calculate(LinkedList<Token> tokenList){
		String result = new String();

	    String unknownToken = new String();
	    for(Token token : tokenList)
	    {
	        if(token.getType() == Token.EType.NoType)
	        {
	            if(unknownToken.length() > 0)
	            {
	                unknownToken += ",";
	            }
	            unknownToken += token.getContent();
	        }
	    }
	    if(unknownToken.length() > 0)
	    {
	        currentRecord.success = false;
	        return systemContext.getString(R.string.error_unknown_keyword) + unknownToken;
	    }
	    
	    BuildContext bContext = new BuildContext(systemContext, constManager, tokenList);

	    Expr expr = AdditiveExpr.buildExpr(bContext);
	    if(expr != null){
	    	if(bContext.tokenList.size() > 0){
	    		String tokenString = new String();
	    		for(Token token : tokenList){
	    			if(tokenString.length() != 0){
	    				tokenString += ",";
	    			}
	    			tokenString += token.getContent();
	    		}
	    		currentRecord.success = false;
    	        result = systemContext.getString(R.string.error_unnecessary_keyword) + tokenString;
	    	}else{
		        EvaluateContext eContext = new EvaluateContext(systemContext, this);
		        eContext.pushResult(new Complex(0));
		        if(expr.evaluate(eContext))
		        {
		        	Complex value = eContext.popResult();
		        	result = eContext.getFormatter().toString(value);
		        	if(result != null){
		        		currentRecord.success = true;
		        		currentRecord.result = value;
		        	}else{
		        		currentRecord.success = false;
		    	        result = systemContext.getString(R.string.error_invalid_input);
		        	}
		        }
		        else
		        {
		            currentRecord.success = false;
		            result = eContext.getErrorMessage();
		        }
	    	}
	    }
	    else
	    {
	        currentRecord.success = false;
	        result = bContext.errorMessage;
	    }
	    return result;
		
	}
	
	boolean isSuccess(){
		return currentRecord.success;
	}
	
	private Record currentRecord;
	private LinkedList<Record> recordList = new LinkedList<Record>();
}

File: Interpreter/CalculateEngine.py
import math
from Interpreter.FunctionManager import *
from Interpreter.ConstManager import *
from Interpreter.R_string import *
from Interpreter.Complex import *
from Interpreter.PreDefineFunction import *
from Interpreter.TokenPatternFactory import *
from Interpreter.PatternBuilder import *
from Interpreter.TokenPattern import *
from Interpreter.TokenAnalyzer import *

#计算引擎
class CalculateEngine:
	CUSTOM_FUN_COUNT = 12

	class Record:
		def __init__(self):
			self.success = None
			self.question = None
			self.result = None

	def __init__(self, context):
		self.systemContext = context
		self.functionManager = FunctionManager(context)
		self.constManager = ConstManager()
		self.currentRecord = None
		self.recordList = []
		self.patterns = []
		self.build_basic_patterns()
	# 创建四则运算规则
	def build_basic_patterns(self):
		numberPattern = r"(((\.[0-9]+)|([0-9]+(\.[0-9]*)?))[eE][+-]?[0-9]+)"
		numberPattern = numberPattern + "|"
		numberPattern = numberPattern + r"((\.[0-9]+)|([0-9]+\.[0-9]*))"
		numberPattern = numberPattern + "|"
		numberPattern = numberPattern + "([0-9]+)"
		self.patterns.append(TokenPattern(TokenType.Number, "(" + numberPattern + ")"))
		self.patterns.append(TokenPattern(TokenType.Operator, r"[-+*/]"))
		self.patterns.append(TokenPattern(TokenType.Parenthese, "[()]"))
		self.patterns.append(TokenPattern(TokenType.Comma, ","))

	def register_const(self, name, value):
		self.constManager.registerConst(name, value)

	def register_function(self, fun):
		self.functionManager.registerFunction(fun)

	# 计算表达式
	def calculate(self, strQuestion, convertFormat):
		if not self.currentRecord:
			self.currentRecord = self.Record()
		# 替换表达式中的变量名
		strSrc = strQuestion
		strAnalyze = strQuestion
		for k, v in self.systemContext.get_values().items():
			strAnalyze = strSrc.replace(k, v)
			strSrc  = strAnalyze
		# 对表达式进行分析，返回语法树
		tokenList = self.analyzeToken(strAnalyze)
		if not convertFormat:
			self.currentRecord.question = strQuestion
		# 对语法树进行计算，返回运算结果
		return self.calculate1(tokenList)

	# 表达式分析
	def analyzeToken(self, strQuestion):
		# 生成分析语法对象集合
		patterns = []
		fun_pattern = PatternBuilder.build(self.functionManager.functions())
		if len(fun_pattern) > 0:
			#函数语法定义
			patterns.append(TokenPattern(TokenType.FunctionName, fun_pattern))
		#基本语法定义
		patterns.extend(self.patterns)
		# 对表达式进行分析，返回语法树
		return TokenAnalyzer().analyzeToken(strQuestion, patterns)

	def isFunction(self, name):
		return (self.functionManager.getFunction(name) != None)
	
	def getCustomFunctionItems(self):
		items = []
		
		for i in range(1, self.CUSTOM_FUN_COUNT + 1):
			key = "F" + str(i)
			item = None
			udf = self.functionManager.getUserDefineFunction(key)
			if udf:
				item = udf.get_name() + ":" + udf.getExprString()
			else:
				item = key + ":Empty"
			items.append(item)

		return items

	def saveCustomFunction(self, key, funName, funText):
		return self.functionManager.registerUserDefineFunction(key, funName, funText)

	def clearCustomFunctions(self):
		for i in range(1, self.CUSTOM_FUN_COUNT + 1):
			UserDefineFunction.clear(self.systemContext, "F" + str(i))

	def getConstManager(self):
		return self.constManager

	def getConstItems(self):
		count = self.constManager.getConstCount()
		items = []
		
		list = self.constManager.consts()
		
		formatter = StandardFormatter(self.systemContext)
		
		index = 0
		for key in list:
			item = key + ":" + formatter.toString(self.constManager.find(key))
			items.append(item)
			index = index + 1

		return items

	def getRecordCount(self):
		return len(self.recordList)

	def getRecord(self, index):
		if(index >= 0) and (index < len(self.recordList)):
			return self.recordList[index]
		else:
			return None

	def saveRecord(self):
		if(self.currentRecord != None) and (self.currentRecord.success):
			self.recordList.append(self.currentRecord)
			self.currentRecord = None
			return True
		else:
			return False
	
	def getRecordItems(self):
		formator = StandardFormatter(self.systemContext)
		record_count = self.getRecordCount()
		items = []
		for i in range(0, record_count):
			record = self.getRecord(i)
			recordString = formator.toString(record.result) + ":" + record.question
			items.append(recordString)
		return items

	def clearRecord(self, index):
		if(index >= 0) and (index < len(self.recordList)):
			self.recordList.pop(index)
			return True
		else:
			return False
	
	def clearAllRecord(self):
		self.recordList.clear()
		return True
	
	def getSystemContext(self):
		return self.systemContext

	def calculate1(self, tokenList):
		result = ''
		unknownToken = ''
		for token in tokenList:
			if token.getType() == TokenType.NoType:
				if len(unknownToken) > 0:
					unknownToken = unknownToken + ','
				unknownToken += token.getContent()

		if len(unknownToken) > 0:
			self.currentRecord.success = False
			return str(R_string.error_unknown_keyword) + unknownToken

		bContext = BuildContext(self.systemContext, self.constManager, tokenList)
		expr = AdditiveExpr.buildExpr(bContext)
		if expr:
			if len(bContext.tokenList) > 0:
				tokenString = ''
				for token in tokenList:
					if len(tokenString) != 0:
						tokenString += ","
					tokenString += token.getContent()
				self.currentRecord.success = False
				result = R_string.error_unnecessary_keyword + tokenString
			else:
				eContext = EvaluateContext(self.systemContext, self)
				eContext.pushResult(Complex(0))
				if expr.evaluate(eContext):
					value = eContext.popResult()
					result = eContext.getFormatter().toString(value)
					if result:
						self.currentRecord.success = True
						self.currentRecord.result = value
					else:
						self.currentRecord.success = False
						result = R_string.error_invalid_input
				else:
					self.currentRecord.success = False
					result = eContext.getErrorMessage()
		else:
			self.currentRecord.success = False
			result = bContext.errorMessage
		return result

	def isSuccess(self):
		return self.currentRecord.success
